File: Python DSA/DynamicProgramming/climbing_stairs.py
"""
CLIMBING STAIRS - Easy
Count ways to climb n stairs (1 or 2 steps at a time).

Logical Thinking:
1. Base: 1 way for 1 stair, 2 ways for 2 stairs
2. Current ways = ways[i-1] + ways[i-2]
3. Fibonacci pattern

Logical Thinking:
def climb_stairs(n):
    if n <= 2:
        return n
    prev2, prev1 = 1, 2
    for _ in range(3, n + 1):
        curr = prev1 + prev2
        prev2, prev1 = prev1, curr
    return prev1
"""

# TODO: Implement solution


# Test cases
if __name__ == "__main__":
    print(f"Testing Climbing Stairs...")
    print("Add your test cases here")

File: DSA/DP/Java/ClimbingStairs.java
// LeetCode 70: Climbing Stairs
// You are climbing a staircase. It takes n steps to reach the top.

public class ClimbingStairs {
    public int climbStairs(int n) {
        if (n <= 2) {
            return n;
        }
        
        int prev2 = 1; // f(1)
        int prev1 = 2; // f(2)
        
        for (int i = 3; i <= n; i++) {
            int current = prev1 + prev2;
            prev2 = prev1;
            prev1 = current;
        }
        
        return prev1;
    }
    
    // Recursive approach with memoization
    public int climbStairsMemo(int n) {
        int[] memo = new int[n + 1];
        return climbStairsHelper(n, memo);
    }
    
    private int climbStairsHelper(int n, int[] memo) {
        if (n <= 2) {
            return n;
        }
        
        if (memo[n] != 0) {
            return memo[n];
        }
        
        memo[n] = climbStairsHelper(n - 1, memo) + climbStairsHelper(n - 2, memo);
        return memo[n];
    }
    
    public static void main(String[] args) {
        ClimbingStairs solution = new ClimbingStairs();
        
        int[] testCases = {2, 3, 4, 5, 10};
        
        for (int n : testCases) {
            System.out.println("Ways to climb " + n + " stairs: " + solution.climbStairs(n));
        }
    }
}